Adding user model and userProfile app

yulisuliman opened the pull request:
- creating users app and user model.
using abstractUser- a full User model,as an abstract class so that you can inherit from it and add your own profile fields and methods.
- Adding tests.

fix #20

OmerCS8 reviewed (dismissed):
Looks good, great job!

AlonShakaroff reviewed (changes requested):
Hey, the ResuMeERDiagram.pdf shouldn't be part of this PR.
Other then that everything looks good

MatanP12 reviewed (dismissed):
LGTM

TomerNewmanPrograms reviewed (dismissed):
Looks good to me

AlonShakaroff reviewed (approved):
LGTM

natifridman commented on userProfile/models.py:
```diff
@@ -0,0 +1,16 @@
+
+class userProfile(AbstractUser):
+    profile_pic = models.ImageField(default='profile-icon.png', upload_to='profiles/')
+    email = models.CharField(max_length=30, primary_key=True, null=False, unique=True, validators=[validate_email_addr])
```
Email is already part of django authentication.

natifridman commented on userProfile/models.py:
```diff
@@ -0,0 +1,16 @@
+        raise ValidationError('Email is not valid')
+
+
+class userProfile(AbstractUser):
```
You don't need to create a new user, you can use the django authentication.
You can use [this](https://simpleisbetterthancomplex.com/tutorial/2016/07/22/how-to-extend-django-user-model.html#onetoone) as a reference

natifridman replied:
Please follow pep8 for class names guidelines

natifridman commented on userProfile/migrations/0002_remove_userprofile_id_userprofile_profile_pic_and_more.py:
```diff
@@ -0,0 +1,27 @@
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
```
Can you use a shorter and more meaningful file name? Why do you need 2 migration files?

natifridman commented on userProfile/models.py.bak:
```diff
@@ -0,0 +1,8 @@
+from django.db import models
```
Please remove backup files

natifridman commented on userProfile/models.py:
```diff
@@ -0,0 +1,16 @@
+from django.core.exceptions import ValidationError
+
+
+def validate_email_addr(test_email):
```
You have this function as part of django validators in `django.core.validators`

yulisuliman commented on Pipfile:
```diff
@@ -5,11 +5,13 @@ name = "pypi"
 
 [packages]
 django = "*"
+pillow = "*"
```
Hi, I need this only in [packages] , will remove it in the other package.
I need pillow for saving images in database( for user profile picture).

MatanP12 reviewed (approved):
LGTM

natifridman commented on userProfile/apps.py:
```diff
@@ -0,0 +1,10 @@
+from django.apps import AppConfig
+
+
+class UserprofileConfig(AppConfig):
```
Please follow [pep 8](https://peps.python.org/pep-0008/#class-names) for naming convention. Also for the [app naming convention](https://learndjango.com/tutorials/django-best-practices-projects-vs-apps), maybe you can call it `user`?

AlonShakaroff replied:
@natifridman Maybe it should be `users` because according to the name convention it should be the plural of the main model.

natifridman replied:
That's correct

natifridman commented on userProfile/signals.py:
```diff
@@ -0,0 +1,15 @@
+
+
+@receiver(post_save, sender=User)
+def create_profile(sender, instance, created, **kwargs):
```
Why does this need to be in a different file than the model? Something like [this](https://github.com/redhat-beyond/falcon/blob/main/users/models.py)

OmerCS8 replied:
In Django's documentation they recommended adding signals in this way.
It's supposed to prevent import problems.
![SmartSelect_20220328-160347_Chrome.jpg](https://user-images.githubusercontent.com/96747472/160403814-8f5ae808-2fe6-4c25-b422-6d812e4d1a34.jpg)
https://docs.djangoproject.com/en/4.0/topics/signals/

OmerCS8 reviewed (approved):
LGTM

AlonShakaroff reviewed (approved):
Looks good to me!

MatanP12 reviewed (dismissed):
LGTM!

natifridman commented on users/tests.py:
```diff
@@ -0,0 +1,50 @@
+
+@pytest.mark.django_db()
+def test_new_user(new_user):
+    assert new_user.username == USERNAME
+    assert new_user.first_name == FIRSTNAME
+    assert new_user.last_name == LASTNAME
+    assert new_user.password == PASSWORD
+    assert new_user.email == EMAIL
```
Why do you need to check this?

yulisuliman replied:
To validate that our user values are the same as the user inputs

natifridman reviewed (commented):
Nice work! add some comments in the tests.
Maybe we can add some more tests? For example, invalid values for users

yulisuliman commented on users/tests.py:
```diff
@@ -0,0 +1,50 @@
+
+
+@pytest.mark.django_db()
+def test_delete_user(new_user):
```
The difference is that when we delete a user we expect that both the user and the profile will be deleted (Delete User test) , and when we delete only the profile we expect that only the profile will be deleted and not the user itself (Delete profile test).

yulisuliman commented on users/tests.py:
```diff
@@ -0,0 +1,50 @@
+
+
+@pytest.mark.django_db()
+def test_persist_user(new_user):
```
In this test we check if the user is saved in the database.
In the new user test we check that the user values are the same as the user inputs.

abiton1 commented on ResuMe/settings.py:
```diff
@@ -28,6 +28,8 @@
 ALLOWED_HOSTS = []
 
 
+ALLOWED_HOSTS = ['*']
```
Duplicate deceleration of `ALLOWED_HOSTS`

OmerCS8 reviewed (dismissed):
LGTM

OmerCS8 reviewed (approved):
LGTM

yulisuliman commented:
Hi,
I set the Django User email as unique.
As Nati suggested I added some tests to validate unique email and username.
for some reason the tests are pass locally on my machine, but not on the GitHub actions.
For now I deleted the unique email test form the tests.py files.
I will add this test later on, please review my PR so we could continue working on the other models.

MatanP12 reviewed (dismissed):
LGTM

TomerNewmanPrograms reviewed (approved):
looks very good

AlonShakaroff reviewed (dismissed):
Looks good

natifridman commented on users/tests.py:
```diff
@@ -0,0 +1,97 @@
+# Check invalid email
+@pytest.mark.django_db
+def test_validate_email_addr():
+    try:
```
You can expect for exception with [pytest.raises](https://docs.pytest.org/en/6.2.x/assert.html#assertions-about-expected-exceptions).
So in this case we won't need to use try and not even assert

MatanP12 reviewed (approved):
LGTM

TomerNewmanPrograms reviewed (dismissed):
Lgtm

OmerCS8 reviewed (approved):
Great job!

AlonShakaroff reviewed (dismissed):
LGTM

natifridman commented:
@yulisuliman Can you please squash your commits so we can merge the PR?

MatanP12 reviewed (approved):
LGTM

TomerNewmanPrograms reviewed (approved):
Great!

OmerCS8 reviewed (approved):
LGTM
